### PR TITLE
Implement Client::PatchBucket.

### LINKED ISSUE #823: Implement Client::PatchBucketMetadata
This is a wrapper around [Buckets: patch](https://cloud.google.com/storage/docs/json_api/v1/buckets/patch).  The member function prototype should be similar to:

```C++
class Client { public:
BucketMetadata PatchBucketMetadata(
    BucketMetadata original_metadata,
    BucketMetadata new_metadata,
    Parameters&&... parameters);
};
```

This task is completed when:

- [ ] The new API has been added to Client.
- [ ] The new API has a completed doxygen comment block.
- [ ] There is a unit test for this API that checks the function works in a easy case.
- [ ] There is a unit test for this API that checks retries work correctly.
- [ ] There is a unit test for this API that checks errors are reported correctly.
- [ ] The uni tests run in the CI builds.
- [ ] There is an integration test for this API that runs against the emulator and against the GCS service.
- [ ] There is a code snippet showing how to use the API.
- [ ] The code snippet and the integration tests run in the CI builds.

### PR DESCRIPTION
**Pull request by coryan**
This fixes #823, and computes the Bucket CRUD stuff. It implements the
API, the unit tests, an integration test, code snippets, and runs the
snippets in the CI build.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1001)

### CODE REVIEW

File: google/cloud/storage/tests/bucket_integration_test.cc [hunk 1]
```diff
@@ -113,6 +113,28 @@ TEST_F(BucketIntegrationTest, BasicCRUD) {
...
+                         LifecycleRule::MaxAge(30),
+                         LifecycleRule::MatchesStorageClassStandard()),
+                     LifecycleRule::Delete());
+  desired_state.set_storage_class(storage_class::Standard())
+      .set_lifecycle(BucketLifecycle{{rule}})
+      .set_website(BucketWebsite{"index.html", "404.html"});
+
+  BucketMetadata patched =
+      client.PatchBucket(bucket_name, updated_meta, desired_state);
+  EXPECT_EQ(storage_class::Standard(), patched.storage_class());
```
Comment (houglum) [inline google/cloud/storage/tests/bucket_integration_test.cc]: Seems like we ought to have: 
1) A test like this that only sets some fields
2) A more exhaustive test that alters all the fields and makes sure the resulting bucket has all the correct attributes (i.e. that the PATCH body was formed correctly and the given request acts as expected).

Thoughts?

File: google/cloud/storage/examples/storage_bucket_samples.cc [hunk 2]
```diff
@@ -193,6 +237,8 @@ int main(int argc, char* argv[]) try {
       {"get-bucket-metadata", &GetBucketMetadata},
       {"delete-bucket", &DeleteBucket},
       {"change-default-storage-class", &ChangeDefaultStorageClass},
+      {"patch-bucket", &PatchBucket},
```
Comment (frankyn) [inline google/cloud/storage/examples/storage_bucket_samples.cc]: This isn't a generic Bucket patch operation. How about`patch-bucket-storage-class`?

Comment (coryan) [reply google/cloud/storage/examples/storage_bucket_samples.cc]: Done. But note that many of the other examples are "not a generic X operation" either. We have `create-bucket`, but it creates with defaults, and `insert-object` doesn't do anything too fancy, these are examples, not a replacement for `gsutil` (thankfully).

Happy to revert if that changes your mind, or to leave the new changes if it does not change your mind.

**Comment by coryan** [comment]
PTAL.

File: google/cloud/storage/tests/bucket_integration_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/tests/bucket_integration_test.cc]: I created a larger test, which found some bugs :tada:

**Pull request merged**